### PR TITLE
[CORDA-1605]: When one node in a transaction fails to respond, operator of initiating node needs to understand which node it is waiting for (fix)

### PR DESCRIPTION
**Pull request by sollecitom**
https://r3-cev.atlassian.net/browse/CORDA-1605

### CODE REVIEW

File: node/src/main/kotlin/net/corda/node/services/config/NodeConfiguration.kt [hunk 1]
```diff
@@ -469,3 +477,21 @@ data class SecurityConfiguration(val authService: SecurityConfiguration.AuthServ
         }
     }
 }
+
+interface A {
```
Comment (exFalso) [inline node/src/main/kotlin/net/corda/node/services/config/NodeConfiguration.kt]: Remove this xD

File: node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+import java.util.concurrent.ScheduledExecutorService
+import java.util.concurrent.TimeUnit
+
+internal class FlowMonitor private constructor(private val retrieveFlows: () -> Set<FlowStateMachineImpl<*>>,
+                                               private val monitoringPeriod: Duration,
+                                               private val suspensionLoggingThreshold: Duration,
+                                               private val scheduler: ScheduledExecutorService,
+                                               private val shutdownScheduler: Boolean) : LifecycleSupport {
+
+    internal constructor(retrieveFlows: () -> Set<FlowStateMachineImpl<*>>, monitoringPeriod: Duration, suspensionLoggingThreshold: Duration, scheduler: ScheduledExecutorService) : this(retrieveFlows, monitoringPeriod, suspensionLoggingThreshold, scheduler, false)
```
Comment (exFalso) [inline node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt]: These lines are too long

File: node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+    internal constructor(retrieveFlows: () -> Set<FlowStateMachineImpl<*>>, monitoringPeriod: Duration, suspensionLoggingThreshold: Duration) : this(retrieveFlows, monitoringPeriod, suspensionLoggingThreshold, defaultScheduler(), true)
+
+    private companion object {
+        private fun defaultScheduler() = Executors.newSingleThreadScheduledExecutor()
```
Comment (exFalso) [inline node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt]: This starts the thread already for some reason. I usually put these in a nullable var and only create them on start()

File: node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        synchronized(started) {
+            if (shutdownScheduler) {
+                scheduler.shutdown()
+            }
+            started = false
+        }
+    }
+
+    private fun logFlowsWaitingForParty(suspensionLoggingThreshold: Duration) {
+        val now = now()
```
Comment (exFalso) [inline node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt]: Perhaps use `Instant.now()` it's easier to read code if the symbol scope is clear from the syntax

File: node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+                    is FlowIORequest.ExecuteAsyncOperation -> "for an asynchronous operation to complete"
+                }
+        )
+        message.append(".")
+        return message.toString()
+    }
+
+    private fun Iterable<FlowSession>.partiesInvolved() = map { it.counterparty }.joinToString(", ", "[", "]")
+
+    private fun FlowStateMachineImpl<*>.ioRequest() = (transientState?.value?.checkpoint?.flowState as? FlowState.Started)?.flowIORequest
```
Comment (exFalso) [inline node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt]: You can also just call `snapshot()` on FlowStateMachine

File: node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        message.append(
+                when (request) {
+                    is FlowIORequest.Send -> "to send a message to parties ${request.sessionToMessage.keys.partiesInvolved()}"
+                    is FlowIORequest.Receive -> "to receive messages from parties ${request.sessions.partiesInvolved()}"
+                    is FlowIORequest.SendAndReceive -> "to send and receive messages from parties ${request.sessionToMessage.keys.partiesInvolved()}"
+                    is FlowIORequest.WaitForLedgerCommit -> "for the ledger to commit transaction with hash ${request.hash}"
+                    is FlowIORequest.GetFlowInfo -> "to get flow information from parties ${request.sessions.partiesInvolved()}"
+                    is FlowIORequest.Sleep -> "to wake up from sleep ending at ${LocalDateTime.ofInstant(request.wakeUpAfter, ZoneId.systemDefault())}"
+                    FlowIORequest.WaitForSessionConfirmations -> "for sessions to be confirmed"
+                    is FlowIORequest.ExecuteAsyncOperation -> "for an asynchronous operation to complete"
```
Comment (exFalso) [inline node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt]: Perhaps include the operation's `.javaClass` in the message as this is a generic operation

File: node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManager.kt [hunk 7]
```diff
@@ -86,9 +86,11 @@ interface StateMachineManager {
     val flowHospital: StaffedFlowHospital
 }
 
-// These must be idempotent! A later failure in the state transition may error the flow state, and a replay may call
-// these functions again
 interface StateMachineManagerInternal {
+    fun snapshot(): Set<FlowStateMachineImpl<*>>
```
Comment (exFalso) [inline node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManager.kt]: This should be on `StateMachineManager`, this interface is used by flows internally. Perhaps better kdocs are due

File: node/src/main/kotlin/net/corda/node/services/config/NodeConfiguration.kt (hunk 1 above)
Comment (sollecitom) [reply node/src/main/kotlin/net/corda/node/services/config/NodeConfiguration.kt]: Haha, teaching Kotlin :)

File: node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt (hunk 3 above)
Comment (sollecitom) [reply node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt]: Refactored to accept an optional ``ScheduledExecutorService`` instead.

**Pull request merged**